[WebXR] Fix MESSAGE_CHECK failure in PlatformXRSystem::RequestFrame()

achan00 opened the pull request:
#### a3fcf4fdfbc19460ad56323d2c0dbb4c26906d2f
<pre>
[WebXR] Fix MESSAGE_CHECK failure in PlatformXRSystem::RequestFrame()
<a href="https://bugs.webkit.org/show_bug.cgi?id=275912">https://bugs.webkit.org/show_bug.cgi?id=275912</a>
<a href="https://rdar.apple.com/130300869">rdar://130300869</a>

Reviewed by Mike Wyrzykowski and Dan Glastonbury.

If the WebXR immersive session was triggered by the system, PlatformXRSystem::sessionDidEnd()
would be called on the UI process side first, which updated the session state to Idle.
However, during the short period when the UI process had handled the session end request but
the web process had not handled it yet, PlatformXRSystem::requestFrame() could still be called
on the UI process side, which would fail the m_immersiveSessionState == SessionRunning check.

To fix this, we introduce two new immersive session states: SessionEndingFromSystem and
SessionEndingFromWebContent. If PlatformXRSystem::sessionDidEnd() is called while the state
is SessionRunning, that means the session end request comes from the system and we&apos;ll update
the state to SessionEndingFromSystem. The message check in PlatformXRSystem::requestFrame()
has been updated to allow the state to be either SessionRunning and SessionEndingFromSystem,
but will skip the actual frame update if it&apos;s SessionEndingFromSystem. When the web process
has ended the XR session, it&apos;ll send the PlatformXRSystem::didCompleteShutdownTriggeredBySystem
message which will reset PlatformXRSystem::m_immersiveSessionState back to Idle.

If the session end request comes from XRSession.end(), then PlatformXRSystem::shutDownTrackingAndRendering()
will be called first when the state is SessionRunning, and the session state will be updated
to SessionEndingFromWebContent. After the system has finished the work to end the session,
PlatformXRSystem::sessionDidEnd() will be called which will update the session state to Idle.

* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::shutdown):
If the ending of the XR session is triggered by the system, call
PlatformXR::Device::didCompleteShutdownTriggeredBySystem() after
finishing the shutdown work.
* Source/WebCore/platform/xr/PlatformXR.h:
(PlatformXR::Device::didCompleteShutdownTriggeredBySystem):
* Source/WebKit/Shared/XR/XRDeviceProxy.cpp:
(WebKit::XRDeviceProxy::didCompleteShutdownTriggeredBySystem):
(WebKit::XRDeviceProxy::requestFrame):
Make sure the completion handler is always called.
* Source/WebKit/Shared/XR/XRDeviceProxy.h:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::shutDownTrackingAndRendering):
(WebKit::PlatformXRSystem::requestFrame):
Allow the session state to be either SessionRunning and SessionEndingFromSystem.
But skip the frame update if it&apos;s SessionEndingFromSystem.
Make sure the completion handler is called even if there&apos;s no xrCoordinator.
(WebKit::PlatformXRSystem::submitFrame):
Allow the session state to be either SessionRunning and SessionEndingFromSystem.
(WebKit::PlatformXRSystem::didCompleteShutdownTriggeredBySystem):
Make sure we&apos;ll only get here from handling a session end request from the system.
Reset the state back to Idle.
(WebKit::PlatformXRSystem::sessionDidEnd):
If this is called when the state is SessionRunning, the session is being ended
by the system so set the next state to be SessionEndingFromSystem. Otherwise,
reset it back to Idle.
(WebKit::PlatformXRSystem::invalidateImmersiveSessionState):
* Source/WebKit/UIProcess/XR/PlatformXRSystem.h:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.messages.in:
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.cpp:
(WebKit::PlatformXRSystemProxy::didCompleteShutdownTriggeredBySystem):
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.h:

Canonical link: <a href="https://commits.webkit.org/280392@main">https://commits.webkit.org/280392@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a0a94f420f8d3c0a0fde6ae364ec19228927704

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56487 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35813 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8959 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60094 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6923 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58613 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43435 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7119 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/45760 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4846 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58516 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33676 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48746 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26621 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30453 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6071 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/5927 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6343 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61778 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/395 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6459 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/53015 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/395 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48807 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/52921 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12509 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/350 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/31640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32726 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33809 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32473 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (approved):
Thanks for fixing this!

djg reviewed (approved):
:shipit:

achan00 committed:
[WebXR] Fix MESSAGE_CHECK failure in PlatformXRSystem::RequestFrame()
https://bugs.webkit.org/show_bug.cgi?id=275912
rdar://130300869

Reviewed by Mike Wyrzykowski and Dan Glastonbury.

If the WebXR immersive session was triggered by the system, PlatformXRSystem::sessionDidEnd()
would be called on the UI process side first, which updated the session state to Idle.
However, during the short period when the UI process had handled the session end request but
the web process had not handled it yet, PlatformXRSystem::requestFrame() could still be called
on the UI process side, which would fail the m_immersiveSessionState == SessionRunning check.

To fix this, we introduce two new immersive session states: SessionEndingFromSystem and
SessionEndingFromWebContent. If PlatformXRSystem::sessionDidEnd() is called while the state
is SessionRunning, that means the session end request comes from the system and we'll update
the state to SessionEndingFromSystem. The message check in PlatformXRSystem::requestFrame()
has been updated to allow the state to be either SessionRunning and SessionEndingFromSystem,
but will skip the actual frame update if it's SessionEndingFromSystem. When the web process
has ended the XR session, it'll send the PlatformXRSystem::didCompleteShutdownTriggeredBySystem
message which will reset PlatformXRSystem::m_immersiveSessionState back to Idle.

If the session end request comes from XRSession.end(), then PlatformXRSystem::shutDownTrackingAndRendering()
will be called first when the state is SessionRunning, and the session state will be updated
to SessionEndingFromWebContent. After the system has finished the work to end the session,
PlatformXRSystem::sessionDidEnd() will be called which will update the session state to Idle.

* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::shutdown):
If the ending of the XR session is triggered by the system, call
PlatformXR::Device::didCompleteShutdownTriggeredBySystem() after
finishing the shutdown work.
* Source/WebCore/platform/xr/PlatformXR.h:
(PlatformXR::Device::didCompleteShutdownTriggeredBySystem):
* Source/WebKit/Shared/XR/XRDeviceProxy.cpp:
(WebKit::XRDeviceProxy::didCompleteShutdownTriggeredBySystem):
(WebKit::XRDeviceProxy::requestFrame):
Make sure the completion handler is always called.
* Source/WebKit/Shared/XR/XRDeviceProxy.h:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::shutDownTrackingAndRendering):
(WebKit::PlatformXRSystem::requestFrame):
Allow the session state to be either SessionRunning and SessionEndingFromSystem.
But skip the frame update if it's SessionEndingFromSystem.
Make sure the completion handler is called even if there's no xrCoordinator.
(WebKit::PlatformXRSystem::submitFrame):
Allow the session state to be either SessionRunning and SessionEndingFromSystem.
(WebKit::PlatformXRSystem::didCompleteShutdownTriggeredBySystem):
Make sure we'll only get here from handling a session end request from the system.
Reset the state back to Idle.
(WebKit::PlatformXRSystem::sessionDidEnd):
If this is called when the state is SessionRunning, the session is being ended
by the system so set the next state to be SessionEndingFromSystem. Otherwise,
reset it back to Idle.
(WebKit::PlatformXRSystem::invalidateImmersiveSessionState):
* Source/WebKit/UIProcess/XR/PlatformXRSystem.h:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.messages.in:
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.cpp:
(WebKit::PlatformXRSystemProxy::didCompleteShutdownTriggeredBySystem):
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.h:

Canonical link: https://commits.webkit.org/280392@main
